cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
     }
 }
 
+/// Get the number of general-purpose and floating-point registers required to
+/// hold the given ABI parameters.
+fn num_registers_required<'a>(
+    word_bit_size: u8,
+    call_conv: CallConv,
+    shared_flags: &shared_settings::Flags,
+    isa_flags: &isa_settings::Flags,
+    params: impl IntoIterator<Item = &'a AbiParam>,
+) -> (usize, usize) {
+    let gprs = &[RU::rax; 128];
+    let mut assigner = Args::new(
+        word_bit_size,
+        gprs,
+        std::usize::MAX,
+        call_conv,
+        shared_flags,
+        isa_flags,
+    );
+
+    let mut gprs_required = 0;
+    let mut fprs_required = 0;
+
+    for p in params {
+        // We're going to mutate the type as it gets converted, so make our own
+        // copy that isn't visible to the outside world.
+        let mut p = p.clone();
+
+        let mut split_factor = 1;
+
+        'converting_param: loop {
+            match assigner.assign(&p) {
+                ArgAction::Convert(ValueConversion::IntSplit) => {
+                    split_factor *= 2;
+                    p.value_type = p.value_type.half_width().unwrap();
+                }
+                ArgAction::Convert(ValueConversion::VectorSplit) => {
+                    split_factor *= 2;
+                    p.value_type = p.value_type.half_vector().unwrap();
+                }
+                ArgAction::Assign(ArgumentLoc::Reg(_))
+                | ArgAction::Convert(ValueConversion::IntBits)
+                | ArgAction::Convert(ValueConversion::Sext(_))
+                | ArgAction::Convert(ValueConversion::Uext(_)) => {
+                    // Ok! We can fit this (potentially split) value into a
+                    // register! Add the number of params we split the parameter
+                    // into to our current counts.
+                    if p.value_type.is_float() {
+                        fprs_required += split_factor;
+                    } else {
+                        gprs_required += split_factor;
+                    }
+
+                    // But we also have to call `assign` once for each split value, to
+                    // update `assigner`'s internal state.
+                    'draining_assignments: for _ in 1..split_factor {
+                        match assigner.assign(&p) {
+                            ArgAction::Assign(_)
+                            | ArgAction::Convert(ValueConversion::IntBits)
+                            | ArgAction::Convert(ValueConversion::Sext(_))
+                            | ArgAction::Convert(ValueConversion::Uext(_)) => {
+                                continue 'draining_assignments;
```

Comment:
I find that this is easier to read and understand when there are nested loops, just so that everyone is clear on what control flow is doing and doesn't wonder "did this happen to move and change which thing is being `continue`d or `break`ed?" but if you feel that it doesn't help code clarity I can remove it.

Reply:
For me, seeing these labels rings a bell and tells me that something weird/complex is going on: when looking at this code, when I saw the labels, I imagined that an innermost loop would break out to the outermost loop. It wasn't the case, so I figured these labels were perhaps here to ease reading. So I was looking for complexity that didn't exist, triggered by the loop header labels. I would be in favor of removing them because 1. they may introduce confusion for readers, 2. minimalism in principle. Since it's highly subjective, I'll let you make the final call.

Reply:
Ok, I can get rid of them, because that's not the reaction I want from code readers ;)